### PR TITLE
Implement transient input for some properties

### PR DESCRIPTION
**Pull request by prusse-martin**

### CODE REVIEW

**Comment by prusse-martin** [comment]
@arthursoprana 
@williamjamir 
Can you give your opinion about this?
I am working to get the alfacase <-> casedescripton working first.

**Comment by williamjamir** [comment]
> @arthursoprana
> @williamjamir
> Can you give your opinion about this?
> I am working to get the alfacase <-> casedescripton working first.

For me is okay, the `attrib_curve` is following the same flow as the other types so I have nothing to add.

Only this new parameter `error_context ` let me confused, was the "default" message not good enough? maybe you can add a test to make the intention more explicit.

**Comment by prusse-martin** [comment]
Make breaking change.
Replace `OpeningCurveDescription` for `Curve`.

**Comment by prusse-martin** [comment]
There is a lot of generated code that add considerable size to this PR and may hide other changes.
It does not help that github colapse some of the relevant files.

File: docs/source/alfacase_definitions/LinearIPRDescription.txt [hunk 1]
```diff
@@ -17,6 +19,14 @@
...
             well_index: 
                 value: number
                 unit: string # optional
+            well_index_curve: 
+                image:
+                    values: [number]
+                    unit: string
+                domain:
+                    values: [number]
+                    unit: string # optional
```
Comment (prusse-martin) [inline docs/source/alfacase_definitions/LinearIPRDescription.txt:32]: `well_index_curve` is optional, not  `well_index_curve.domain.unit` another open PR already fix this,

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py [hunk 2]
```diff
@@ -10,10 +10,12 @@
 from typing import List
```
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:10]: Don't skip this file.

File: src/alfasim_sdk/_internal/alfacase/case_description.py [hunk 3]
```diff
@@ -13,9 +13,12 @@
 from attr.validators import in_
```
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/case_description.py:13]: Don't skip this file.

File: src/alfasim_sdk/_internal/alfacase/case_description_attributes.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+import textwrap
 from enum import EnumMeta
```
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/case_description_attributes.py:2]: Don't skip this file.

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py [hunk 5]
```diff
@@ -10,13 +10,15 @@
```
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py:10]: Don't skip this file.

File: src/alfasim_sdk/_internal/alfacase/generate_schema.py [hunk 6]
```diff
@@ -11,6 +11,7 @@
```
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/generate_schema.py:11]: Don't skip this file.

File: tests/alfacase/test_case_to_alfacase.py [hunk 7]
```diff
@@ -44,6 +49,39 @@ def test_convert_description_to_alfacase_with_empty_dict(tmp_path):
...
+            assert has_key_in_yaml("temperature")
+            assert not has_key_in_yaml("temperature_curve")
+        elif input_type == MultiInputType.Curve:
+            assert not has_key_in_yaml("temperature")
+            assert has_key_in_yaml("temperature_curve")
+
+    else:
+        assert has_key_in_yaml("temperature_input_type")
+        assert has_key_in_yaml("temperature")
+        assert has_key_in_yaml("temperature_curve")
```
Comment (nicoddemus) [inline tests/alfacase/test_case_to_alfacase.py:82]: Hmmm we have an empty key in there, or a key with the defaul value?

File: tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt [hunk 8]
```diff
@@ -32,5 +32,11 @@
                 value: number
                 unit: string # optional
             opening_curve_interpolation_type: \ :class:`InterpolationType <alfasim_sdk._internal.constants.InterpolationType>`\ # optional
-            opening_curve: \ :class:`opening_curve_description_schema <OpeningCurveDescription>`\ # optional
+            opening_curve: 
+                image:
+                    values: [number]
```
Comment (nicoddemus) [inline tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt:37]: Is it simple to change this to `[numbers]` (plural)?

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Awesome work! This definitely shows how much work went into this.

Left a few comments.

File: src/alfasim_sdk/_internal/alfacase/case_description_attributes.py (hunk 4 above)
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/case_description_attributes.py:2]: > ?

There are a lot of generated code around and some of the files I did edit manually require reviewers to click "Load diff" because they are not being rendered by default.
Because of that I added those. I think I could have written better.

Comment (nicoddemus) [reply src/alfasim_sdk/_internal/alfacase/case_description_attributes.py:2]: Ahh OK.

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py [hunk 9]
```diff
@@ -77,6 +82,96 @@ def get_category_for(unit: Optional[str]) -> Optional[str]:
...
+    Update the multi input flags in `item_description` if not present in `alfacase_content` and
+    can be unambiguously deduced from the presence of contant and transient inputs.
+
+    :returns: The updated `item_description` (can return `item_description` unchanged.
+    """
+    item_dict = attr.asdict(item_description, recurse=False)
+    fields_to_update: Dict[str, constants.MultiInputType] = {}
+    for key, value in item_dict.items():
+        if (
+            key.endswith(constants.MULTI_INPUT_TYPE_SUFFIX)
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:98]: That function will make the code removing suffixes really better.

About the check.
It is probably enough since that `Enum` has just this purpose, and any same input for this function will abide to that.
I added the extra check to guard against possible errors.
I could remove the suffix check altogether or move it to the if's body as `assert key.endswith(constants.MULTI_INPUT_TYPE_SUFFIX)` for safekeeping. (I am doing that extra suffix check somewhere else I think).

I will move it to the if's body. If you have any objection let me know.

[the remark this answers is not in the document]
Comment (nicoddemus) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:98]: > I added the extra check to guard against possible errors.

But this will actually hide errors, like someone using a different suffix but as a `MultiInput` value... that's the main point of my question, just checking the instance is not enough?

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py [hunk 10]
```diff
@@ -77,6 +82,96 @@ def get_category_for(unit: Optional[str]) -> Optional[str]:
         return UnitDatabase.GetSingleton().GetDefaultCategory(unit)
 
 
+def update_multi_input_flags(
+    alfacase_content: DescriptionDocument, item_description: T
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:86]: I used that for consistency.
Most of loaders use `document` or `alfacase_content`. While checking the code to better reply this comment I realized most time (maybe always) I call this function the current scope variable is `document`.
But `document` can be a bit misleading since it ussually is a fragment (not the full alfacase document).

I will update it to use the same name as the call sites (just `document`).

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py [hunk 11]
```diff
@@ -77,6 +82,96 @@ def get_category_for(unit: Optional[str]) -> Optional[str]:
...
+                getattr(item_description, key) != new_value
+            ):
+                fields_to_update[key] = new_value
+
+    if fields_to_update:
+        item_description = attr.evolve(item_description, **fields_to_update)
+    return item_description
+
+
+def load_instance(alfacase_content: DescriptionDocument, class_: Type[T]) -> T:
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:124]: See my answer to the previous comment: https://github.com/ESSS/alfasim-sdk/pull/148#discussion_r620576451

File: src/alfasim_sdk/_internal/alfacase/generate_schema.py [hunk 12]
```diff
@@ -26,86 +27,98 @@ def is_enum(value):
...
+def dict_to_alfacase_schema(type_, indent):
     key_type, value_type = (
         typing_inspect.get_last_args(type_)
         if sys.version_info.minor == 6
         else typing_inspect.get_args(type_)
     )
-    return f"MapPattern(Str(), {_get_attr_value(value_type)})"
+    body_indent = indent + 1
+    value_schema = _get_attr_value(value_type, indent=body_indent)
+    if "\n" in value_schema:
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/generate_schema.py:95]: =)
It was annoying to see it generating things like:
```python
        Optional("mass_flow_rates_curve"): MapPattern(Str(), Map(
                {
                    "image": Map({"values": Seq(Float()), "unit": Str()}),
                    "domain": Map({"values": Seq(Float()), "unit": Str()}),
                }
            ),),
```

File: tests/alfacase/test_case_to_alfacase.py (hunk 7 above)
Comment (prusse-martin) [reply tests/alfacase/test_case_to_alfacase.py:82]: Actually this section of the test is just to cover this "feature" in a unit test.
It dumps all multi input fields to the `.alfacase` file, the usual is just to dump the currently used.
The full dump of multi input field was implemented to make round trip tests easier.

TL;DR;
This is the info the is present in the case description, because they are optional they also are required to have a default value. If the exported case description have not set those values the default will be present.

File: tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt (hunk 8 above)
Comment (prusse-martin) [reply tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt:37]: Yes, it is easy.
But I read `[number]` as "a list of elements with type number" ("a list of numbers").
The same way I will read `[str]` as "a list of strings".

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py]: I did remove the suffix check from the condition.
But let the check as asserts. If in the future we try to repurpose the Enum we will have a clear indication of code that need to be updated.

File: tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt (hunk 8 above)
Comment (nicoddemus) [reply tests/alfacase/test_generate_case_description_docstring/test_generate_case_description_docstring_ValveEquipmentDescription_.txt:37]: Ahhh I thought it was a placeholder like `<number>`. OK my misunderstanding, please leave as is.

**Review by nicoddemus — APPROVED** [review]

**Review by arthursoprana — APPROVED** [review]

File: docs/source/alfacase_definitions/ValveEquipmentDescription.txt [hunk 13]
```diff
@@ -12,7 +12,7 @@
             opening_type: \ :class:`ValveOpeningType <alfasim_sdk._internal.constants.ValveOpeningType>`\  = ValveOpeningType.ConstantOpening
             opening: \ :class:`Scalar <barril.units.Scalar>`\  = Scalar(100.0, '%', 'dimensionless')
             opening_curve_interpolation_type: \ :class:`InterpolationType <alfasim_sdk._internal.constants.InterpolationType>`\  = InterpolationType.Constant
-            opening_curve: \ :class:`OpeningCurveDescription <OpeningCurveDescription>`\  = OpeningCurveDescription()
+            opening_curve: \ :class:`Curve <barril.curve.curve.Curve>`\  = Curve(-, s)[]
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply docs/source/alfacase_definitions/ValveEquipmentDescription.txt:15]: I think the online doc generation need them committed today.

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py [hunk 14]
```diff
@@ -187,6 +282,62 @@ def get_dict_of_arrays_loader(
     )
 
 
+def load_curve(key: str, alfacase_content: DescriptionDocument, category) -> Curve:
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:282]: I think that is the case.
What are the gains? If some external code import them they will already be importing `_internal` package.

File: src/alfasim_sdk/_internal/alfacase/generate_schema.py [hunk 15]
```diff
@@ -143,31 +157,47 @@ def is_scalar(type_):
...
+def array_to_alfacase_schema(type_, indent):
     return 'Map({"values": Seq(Float()), "unit": Str()})'
 
 
+def is_curve(type_):
+    return inspect.isclass(type_) and issubclass(type_, Curve)
+
+
+def curve_to_alfacase_schema(type_, indent):
+    # sub_block_indentation = block_indentation + INDENTANTION
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/generate_schema.py:177]: My bad, a left over.

File: src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nicoddemus) [reply src/alfasim_sdk/_internal/alfacase/alfacase_to_case.py:282]: Yeah no need, we have a `_internal` module exactly for this reason. 👍

File: src/alfasim_sdk/_internal/alfacase/case_description_attributes.py [hunk 16]
```diff
@@ -84,25 +211,106 @@ def attrib_scalar(
...
+    return attr.ib(default=default, converter=to_array, type=Array, metadata=metadata)
+
+
+def attrib_curve(
+    default: Union[CurveLike, AttrNothingType] = attr.NOTHING,
+    is_optional: bool = False,
+    category: Optional[str] = None,
+) -> attr._make._CountingAttr:
+    """
+    Create a new attr attribute with a converter to Scalar accepting also tuple(value, unit).
```
Comment (oliveira-mauricio) [inline src/alfasim_sdk/_internal/alfacase/case_description_attributes.py:271]:
```suggestion
    Create a new attr attribute with a converter to Curve accepting also tuple(image, domain).
```

**Review by oliveira-mauricio — APPROVED** [review]

File: docs/source/alfacase_definitions/ValveEquipmentDescription.txt [hunk 17]
```diff
@@ -32,5 +32,11 @@
...
                 unit: string # optional
             opening_curve_interpolation_type: \ :class:`InterpolationType <alfasim_sdk._internal.constants.InterpolationType>`\ # optional
-            opening_curve: \ :class:`opening_curve_description_schema <OpeningCurveDescription>`\ # optional
+            opening_curve: 
+                image:
+                    values: [number]
+                    unit: string
+                domain:
+                    values: [number]
+                    unit: string # optional
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply docs/source/alfacase_definitions/ValveEquipmentDescription.txt:41]: =)
It was my first comment when I promoted the draft to a full PR.
https://github.com/ESSS/alfasim-sdk/pull/148#discussion_r619824707

[the remark this answers is not in the document]
Comment (oliveira-mauricio) [reply docs/source/alfacase_definitions/ValveEquipmentDescription.txt:41]: Oh ok, I didn't see that.

**Review by williamjamir — APPROVED** [review]

File: src/alfasim_sdk/__init__.py [hunk 18]
```diff
@@ -324,7 +324,6 @@ def get_alfasim_sdk_api_path():
     "NodeDescription",
     "NumericalOptionsDescription",
     "OpenHoleDescription",
-    "OpeningCurveDescription",
```
[the remark this answers is not in the document]
Comment (prusse-martin) [reply src/alfasim_sdk/__init__.py:327]: Updated the history to mention it.
The `OpeningCurveDescription` was replaced by a `Curve`.
